requests/dialer_swarm.go

```diff
@@ -0,0 +1,216 @@
+/*
+ * Copyright (C) 2020 The "MysteriumNetwork/node" Authors.
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package requests
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net"
+	"strings"
+	"time"
+)
+
+// ErrAllDialsFailed is returned when connecting to a peer has ultimately failed.
+var ErrAllDialsFailed = errors.New("all dials failed")
+
+// DialerSwarm is a dials to multiple addresses in parallel and earliest successful connection wins.
+type DialerSwarm struct {
+	// ResolveContext specifies the dial function for doing custom DNS lookup.
+	// If ResolveContext is nil, then the transport dials using package net.
+	ResolveContext ResolveContext
+
+	dialer *net.Dialer
+}
+
+// NewDialerSwarm creates swarm dialer with default configuration.
+func NewDialerSwarm(srcIP string) *DialerSwarm {
+	return &DialerSwarm{
+		dialer: &net.Dialer{
+			Timeout:   60 * time.Second,
+			KeepAlive: 30 * time.Second,
+			LocalAddr: &net.TCPAddr{IP: net.ParseIP(srcIP)},
+		},
+	}
+}
+
+// DialContext connects to the address on the named network using the provided context.
+func (ds *DialerSwarm) DialContext(ctx context.Context, network, addr string) (net.Conn, error) {
+	if ds.ResolveContext != nil {
+		addrs, err := ds.ResolveContext(ctx, network, addr)
+		if err != nil {
+			return nil, &net.OpError{Op: "dial", Net: network, Source: nil, Addr: nil, Err: err}
+		}
+
+		conn, errDial := ds.dialAddrs(ctx, network, addrs)
+		if errDial != nil {
+			errDial.OriginalAddr = addr
+
+			return nil, errDial
+		}
+
+		return conn, nil
+	}
+
+	return ds.dialer.DialContext(ctx, network, addr)
+}
+
+func (ds *DialerSwarm) dialAddrs(ctx context.Context, network string, addrs []string) (net.Conn, *ErrorSwarmDial) {
+	addrChan := make(chan string, len(addrs))
+	for _, addr := range addrs {
+		addrChan <- addr
+	}
+
+	close(addrChan)
+
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	resultCh := make(chan dialResult)
+	err := &ErrorSwarmDial{}
+
+	var active int
+dialLoop:
+	for addrChan != nil || active > 0 {
```

Comment:
I wish all the strength and courage for the future generations that will be reading this.

Would this not work:

```go

func (ds *DialerSwarm) dialAddrs(ctx context.Context, network string, addrs []string) (net.Conn, error) {
	ctx, cancel := context.WithCancel(ctx)
	defer cancel()

	resultCh := make(chan dialResult)
	for i := range addrs {
		go ds.dialAddr(ctx, network, addrs[i], resultCh)
	}

	var responses int
	for {
		select {
		case v := <-resultCh:
			responses++
			if v.Err != nil || v.Conn == nil {
				if responses == len(addrs) {
					return nil, ErrAllDialsFailed
				}
			} 
			return v.Conn, nil
		case <-ctx.Done():
			return nil, errors.New("cancelled")
		}
	}
}
```

Reply:
It's ideas from `net` and `github.com/libp2p/go-libp2p` packages for better performance.
Anyway simplified for sake of readability.

Reply:
Indeed reverted this change, `resultCh` is left unclosed for every dial